core/src/tpu.rs

```diff
@@ -208,6 +212,22 @@ impl Tpu {
     }
 
     pub fn join(self) -> thread::Result<()> {
+        // spawn a new thread to wait for tpu close
+        let (sender, receiver) = bounded(0);
+        let _ = thread::spawn(move || {
+            let _ = self.do_join();
+            sender.send(()).unwrap();
+        });
+
+        // timeout of 10s for closing tpu
+        let timeout = Duration::from_secs(TPU_THREADS_JOIN_TIMEOUT_SECONDS);
+        if let Err(RecvTimeoutError::Timeout) = receiver.recv_timeout(timeout) {
+            error!("timeout for closing tvu");
```

Comment:
should we just panic here too?  may as well be explicit  (same with the TPU instance)

Reply:
Good question. I had thought about this too and decided to not panic. 

1. when deadlock, panic will still make the test fail, unless we mark panic expected in the test. If we mark panic expected in test, which will mask the actual panic in the test, and prevent us from detecting other deadlock that may have introduced in future. 
2. panic may prevent other services clean code to run. 
3. since we are already shutting down, printing an error message could be enough.